### PR TITLE
DE-3032 Ignore files

### PR DESCRIPTION
**Pull request by VladimirShteyn**
Enable ignoring files for upload using regular expressions. They may be passed in as CLI args:

```shell
ts-sdk put ... --exclude \.venv/ __pycache__
```

using the existing config file format

```json
{
  "exclude": ["\\.venv/", "__pycache__"]
}
```

or using environment variables,

```shell
export TS_EXCLUDE="['\\.venv/', '__pycache__']"
```

Reviewers should start with the `compress_files` function in `cli/__put_cmd.py`.

Currently `FilenameFilter` is handling both config parsing and filtering files. In the future we can consider splitting that functionality.

Question: How do I run the integration tests? They don't pass on my machine. They do pass in CI.

### CODE REVIEW

**Comment by mlipschultz** [comment]
It looks like you missed [`__tests__/.DS_Store`](https://github.com/tetrascience/ts-sdk-python/pull/84/files#diff-0926e16f0390a49b6ec0cf56a4956ebc75bce305c56148020c4f17076b442781)

File: ts_sdk/task/__util_datalake.py [hunk 1]
```diff
@@ -138,7 +140,12 @@ def upload(self):
 
 
 class Datalake:
-    def __init__(self, endpoint):
+    def __init__(self, endpoint: Optional[str]) -> None:
```
Comment (mlipschultz) [inline ts_sdk/task/__util_datalake.py:143]: Thanks for taking the time to fix some broken windows while you were working on this ticket!

**Comment by VladimirShteyn** [comment]
> It looks like you missed:
> 
> * [x]  [`__tests__/.DS_Store`](https://github.com/tetrascience/ts-sdk-python/pull/84/files#diff-0926e16f0390a49b6ec0cf56a4956ebc75bce305c56148020c4f17076b442781)
> * [x]  [`__tests__/unit/snapshots/.DS_Store`](https://github.com/tetrascience/ts-sdk-python/pull/84/files#diff-03466e5881d18c9527351b4e9b705ad89f905ec2887a5f17b610450d5698f305)

My bad. Fixed!

**Review by yandryianets — APPROVED** [review]

**Review by spinwang — CHANGES_REQUESTED** [review]
Maybe I missed something, but the PR seems to be different from the intent of the ticket. According to the PR, the user of the ts-sdk/SSP will have to manually add the CLI arguments such as `--exclude`, and there are no default values; I am also a little worried that using regex can be non-trivial for a data engineer, especially when it comes to escaping special characters.

Michael Lipschultz Alex Clay Vladimir Shteyn I would like to discuss if we can use a .ignore or .tsignore file following the gitignore syntax and add the files that Toby suggested to the default template.

File: __tests__/unit/cli/test_put_cmd_helpers/test_file_filter.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        assert result.get() == {"foo": ["bar", "baz"]}
+
+
+class TestFilenameFilterAddRegex:
+    """Tests for the "add_regex" method of TestFilenameFilter class."""
+
+    @staticmethod
+    def test_simple() -> None:
+        """
+        Assert that a single regular expression can be added to the `FilenameFilter`.
```
Comment (abclay) [inline __tests__/unit/cli/test_put_cmd_helpers/test_file_filter.py:197]: The test here looks like it's not adding a single regex.  It's adding 2 regexes wrapped in a `FilenameFilter`

**Review by abclay — COMMENTED** [review]
I think you have a good bunch of additions here.  It looks like it can be extended to support some of the types of file exclusions that Spin has been talking about.

File: ts_sdk/cli/put_cmd_helpers/zip_archive_files.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
+import pathlib
+from typing import Iterable, Iterator, Set
+
+from ts_sdk.cli.put_cmd_helpers.file_filter import FilenameFilter
+from ts_sdk.cli.put_cmd_helpers.zip_archive_file import ZipArchiveFile
+
+
+class ZipArchiveFiles(Iterable):
```
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply ts_sdk/cli/put_cmd_helpers/zip_archive_files.py:8]: - [ ] ZipArchiveFiles  -> ZipArchiveFileSet

File: ts_sdk/cli/put_cmd_helpers/file_filter.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        """
+        env_var_data = confuse.EnvSource(prefix=prefix, sep=sep, parse_yaml_docs=True)
+        config.set(env_var_data)
+        return config
+
+    def add_regex(self, other: "FilenameFilter") -> "FilenameFilter":
+        """
+        Add `regex` to the list of regex in this instance and return a new instance.
+        """
+        return self.__class__([*other.regex, *self._regex])
```
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply ts_sdk/cli/put_cmd_helpers/file_filter.py:194]: It seemed like a more intuitive API, kind of like how adding two numbers returns a new number. Thoughts?

[the remark this answers is not in the document]
Comment (abclay) [reply ts_sdk/cli/put_cmd_helpers/file_filter.py:194]: That makes sense.  I read the name of the function as `add a regex to this class` not `combine this class with another one and return the result`

File: ts_sdk/cli/put_cmd_helpers/file_filter.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+class FilenameFilter:
+    """A collection of regular expressions ("regex") used to filter out filenames."""
+
+    #: Name referencing this class and its data.
+    ARG_NAME = "exclude"
+    #: The structure of the configuration file for instantiating this class.
+    TEMPLATE = {ARG_NAME: confuse.Sequence(str)}
+    #: Valid file extensions for configuration files.
+    CONFIG_FILE_EXTENSIONS = (".yaml", ".json")
```
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply ts_sdk/cli/put_cmd_helpers/file_filter.py:19]: OBE because config parsing was moved to the `ensure_args` function.

File: ts_sdk/cli/put_cmd_helpers/file_filter.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            )
+        if filename.suffix not in cls.CONFIG_FILE_EXTENSIONS:
+            plural_ending = "" if len(cls.CONFIG_FILE_EXTENSIONS) == 1 else "s"
+            raise ValueError(
+                f"'{str(filename.name)}' has an invalid file extension! Valid file "
+                f"extension{plural_ending}: {', '.join(cls.CONFIG_FILE_EXTENSIONS)} "
+            )
+
+    @classmethod
+    def from_config(cls, config: confuse.Configuration) -> "FileExcludeRegexConfig":
```
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply ts_sdk/cli/put_cmd_helpers/file_filter.py:70]: It should be returning `"FilenameFilter"` in quotes because at the point that the type hints are evaluated, the name `FilenameFilter` has not been defined. That will cause Python to bonk. Since Python 3.7, the quotes can be removed by doing 

```python
from __future__ import annotations
```

Quotes will not be necessary in Python 4.0.

File: ts_sdk/cli/put_cmd_helpers/file_filter.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+            not a command-line argument.
+            2. The command line arguments are of invalid format or type. See
+            :attr:`~ts_sdk.cli.__put_cmd.FilenameFilter.TEMPLATE` for the
+            expected field and their types.
+        """
+        parsed_config = config.get(cls.TEMPLATE)
+        return cls(parsed_config[cls.ARG_NAME])
+
+    @staticmethod
+    def _update_config_from_yaml(
```
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply ts_sdk/cli/put_cmd_helpers/file_filter.py:89]: Fixed by moving the "exclude" config and environment variable parsing to the `ensure_args` function.

File: __tests__/unit/cli/test_put_cmd_helpers/test_file_filter.py (hunk 2 above)
Comment (VladimirShteyn) [reply __tests__/unit/cli/test_put_cmd_helpers/test_file_filter.py:197]: Updated to 

```
"""
Assert that adding two `FilenameFilter` results in a `FilenameFilter` with
all regexes.
"""
```

File: ts_sdk/cli/__put_cmd.py [hunk 8]
```diff
@@ -110,6 +155,38 @@ def __validate_ids(args):
     )
 
 
+def compress_files(args: argparse.Namespace, buffer: BinaryIO, compression: int = zipfile.ZIP_DEFLATED) -> None:
+    """
+    Compress the folders and files specified in `args` and add to `zip_buffer`.
```
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply ts_sdk/cli/__put_cmd.py:160]: Fixed

**Comment by VladimirShteyn** [comment]
- [ ] Update README.md with instructions for file exclusion.
- [ ] Allow setting "exclude" in the `pyproject.toml` file under `[tool.ts_sdk]`.
- [ ] Always exclude the `.git` folder (see `ensure_args`) function.

**Comment by VladimirShteyn** [comment]
> Maybe I missed something, but the PR seems to be different from the intent of the ticket. According to the PR, the user of the ts-sdk/SSP will have to manually add the CLI arguments such as `--exclude`, and there are no default values; I am also a little worried that using regex can be non-trivial for a data engineer, especially when it comes to escaping special characters.

According to [this poll of the DI Team on Slack](https://tetrascience.slack.com/archives/GMTG3FDNH/p1646773373718019) looks like one third of the DI Team who responded to the poll want glob, another third want regex, and another third don't care. Since regular expressions are already implemented in this PR, let's wait for user feedback before switching to glob. Thoughts?

File: ts_sdk/cli/__put_cmd.py [hunk 9]
```diff
@@ -23,6 +30,20 @@ def put_cmd_args(parser: argparse.ArgumentParser):
     parser.add_argument('version', type=__version_type)
     parser.add_argument('folder', type=__folder_type,
                         help='path to folder to be uploaded')
+    parser.add_argument(
+        f"--{FilenameFilter.ARG_NAME}",
```
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply ts_sdk/cli/__put_cmd.py]: Fixed.

File: README.md [hunk 10]
```diff
@@ -22,30 +22,50 @@ pipenv run pytest
...
+    ".*/__pycache__/.*",
+    ".*/example-input/.*",
+    ".*/example-output/.*",
+    ".*/example-files/.*",
+    ".*/raw_data/.*",
+    ".*/_{0,2}tests?_{0,2}/.*"
+]
+```
+
+In all cases except the `ignore_ssl` and the `exclude` args, environment variable values will be overridden by the configuration file, which are overridden by values passed in via command line. All values set by `exclude` will be used such that setting `"exclude": ["bar"]` in a config file and `export TS_EXCLUDE="baz qux"` would be the equivalent to passing in `--exclude bar baz qux`. If any configuration file or environment variable sets `ignore_ssl` to `True`, its value will be `True` even if `--ignore-ssl false` is passed in.
```
Comment (abclay) [inline README.md:60]: I would maybe reword this to something like:

```
In general, environment variable values will be overridden by the configuration file, which are overridden by values passed in via command line.

Exceptions are:
- `exclude` settings are aggregated: `"exclude": ["bar"]` in a config file and `export TS_EXCLUDE="baz qux"` would be the equivalent to passing in `--exclude bar baz qux`
- If any configuration file or environment variable sets `ignore_ssl` to `True`, its value will be `True` even if `--ignore-ssl false` is passed in.
```

**Review by abclay — APPROVED** [review]
Just one comment about wording in the readme.  Otherwise, looks good to me!

File: README.md (hunk 10 above)
Comment (mlipschultz) [reply README.md:60]: I like Alex's suggestion.  Vlad's proposed change made me initially think there was no environment variable for `ignore_ssl` (although maybe because there wasn't one in the original README)

I wonder why `ignore-ssl` has a different override pattern.

File: README.md [hunk 11]
```diff
@@ -22,30 +22,50 @@ pipenv run pytest
...
-    "auth_token": "your-token",
-    "org": "your-org",
-    "ignore_ssl": false
+  "org": "my-org-slug"
 }
 ```
 
+3. Passing them in as command line arguments.
+
+The `exclude` arg may also be set in your repo's `pyproject.toml` file, under `[tool.ts_sdk]`. For example, to exclude the `.venv` folder, do
```
Comment (mlipschultz) [inline README.md:38]: Is `exclude` too generic?  I'm wondering if we may want to exclude different kinds of things in the future (e.g. vendored dependencies in nodejs task scripts, or even other cli tools may want to exclude things), so having a `TS_EXCLUDE` environment variable could be confusing/abiguous for customers.  It also not as self-documenting as it could be (e.g. it doesn't explain what's being excluded or why).  Perhaps `exclude-paths` or `exclude_path_pattern` would be better?

**Comment by VladimirShteyn** [comment]
@spinwang I followed up with the DI Team on Slack ([link](https://tetrascience.slack.com/archives/C9WD8SHGQ/p1646859975689879)) and looks like the preferred choice is glob. I suggest setting the default values of `--exclude` to something sensible, e.g. [the default excludes used by the builder](https://github.com/tetrascience/ts-lib-artifact-builder/blob/master/builders/task-script.builder.js#L199) and changing the filtering from regex to glob. With those changes implemented, would I get approval to merge the PR?

File: README.md (hunk 11 above)
Comment (VladimirShteyn) [reply README.md:38]: I like this! `--exclude-paths` is way better.

File: README.md (hunk 10 above)
Comment (VladimirShteyn) [reply README.md:60]: - [] Update 

> @spinwang I followed up with the DI Team on Slack ([link](https://tetrascience.slack.com/archives/C9WD8SHGQ/p1646859975689879)) and looks like the preferred choice of customers is glob over regex. I suggest setting the default values of `--exclude` to something sensible, e.g. [the default excludes used by the builder](https://github.com/tetrascience/ts-lib-artifact-builder/blob/master/builders/task-script.builder.js#L199), and changing the filtering from regex to glob. With those changes implemented, would I get approval to merge the PR?

Following up from a chat with @mlipschultz and a separate one from @abclay, looks like this PR is on hold until all stakeholders finalize an implementation.
